### DISCUSSION TITLE
Create ComposableExtensions and ComposableUIKit products

### DISCUSSION THREAD

**Discussion by maximkrouk** [pointfreeco/swift-composable-architecture#587] [Ideas]
Hi, what do you think about creating officially supported libraries for UIKit and general stuff?

UIKit navigation (and behavior in general) is very different from the SwiftUI one, there are some UIKit components, that may be made composable and there are some features for reducers, which can be added as extensions like `onChange` reducer and more.

There is an option to create separate packages, but maybe it will be a little easier to maintain TCA ports if these features will be placed in one package.

Maybe I'll post some examples later and contribute if the idea will fit pointfree's vision of TCA package development path 🙂

**Comment by stephencelis** [comment]
We definitely are interested in exploring this further. Swift 5.5 coming with `async`-`await` hopefully gives us the opportunity to decouple TCA from Combine (and even SwiftUI), and then might be a good time to think about how we could maybe bring UIKit and SwiftUI features out into their own modules.

**Reply by maximkrouk** [reply]
Well, I use TCA with UIKit a lot, so I'll share some solutions in this discussion
> - ComposableViewController ✅
> - ComposableCollectionView ✅
> - ComposableMultisectionCollectionView ❎
> - Composable routing for UIKit ⚠️
> - Composable stuff for ViewControllerLifecycle management ✅
> - Composable stuff for UIKit-compatible modals ✅


> ✅ - Actively using and satisfied with solution
> ❎ - Satisfied with solution, but usage wasn't active
> ⚠️ - Seems to work fine, but the solution is not too neat and may have constraints

Also, my solutions rely on some local stuff you probably not interested in, like:
- https://github.com/MakeupStudio/swift-declarative-configuration (_Framework for closure-based interactions and SwiftUI-like object configuration using `@dynamicMemberLookup` & `callAsFunction`_)
- https://gist.github.com/maximkrouk/35c0ec0baf4d4e797786f60c49a2554e (_The best open-source UIKit button_)

But might be helpful nonetheless

**Comment by maximkrouk** [comment]
Hi once again, I thought about that recently and now I think, that it's actually not a part of the architecture, so that stuff can be done as separate packages, what do you think about creating a community organization like `ComposableArchitectureCommunity` or `TCACommunity` on GitHub?
I already started creating it, but probably it's better if u do that, so if you agree, can u create it and invite me 🙂?
<img width="300" alt="Screenshot 2021-07-22 at 15 54 35" src="https://user-images.githubusercontent.com/40476363/126635106-df4217da-d68e-4b97-b32b-543d7e4224e5.png">

I think a lot of people may be interested in contributing to community solutions for TCA.

**Reply by stephencelis** [reply]
We've had this exact organization in-flight for awhile now but will try to get a lil more momentum behind it. Thanks for the push :smile:
